### PR TITLE
[SPARK-32815][ML] Fix LibSVM data source loading error on file paths with glob metacharacters

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to fix an issue with LibSVM datasource when both of the following are true:
* no user specified schema
* some file paths contain escaped glob metacharacters, such as `[``]`, `{``}`, `*` etc.

The fix is based on another bug fix for CSV/JSON datasources https://github.com/apache/spark/pull/29659.

### Why are the changes needed?
To fix the issue when the follow two queries try to read from paths `[abc]`:
```scala
spark.read.format("libsvm").load("""/tmp/\[abc\].csv""").show
```
but would end up hitting an exception:
```
Path does not exist: file:/private/var/folders/p3/dfs6mf655d7fnjrsjvldh0tc0000gn/T/spark-6ef0ae5e-ff9f-4c4f-9ff4-0db3ee1f6a82/[abc]/part-00000-26406ab9-4e56-45fd-a25a-491c18a05e76-c000.libsvm;
org.apache.spark.sql.AnalysisException: Path does not exist: file:/private/var/folders/p3/dfs6mf655d7fnjrsjvldh0tc0000gn/T/spark-6ef0ae5e-ff9f-4c4f-9ff4-0db3ee1f6a82/[abc]/part-00000-26406ab9-4e56-45fd-a25a-491c18a05e76-c000.libsvm;
	at org.apache.spark.sql.execution.datasources.DataSource$.$anonfun$checkAndGlobPathIfNecessary$3(DataSource.scala:770)
	at org.apache.spark.util.ThreadUtils$.$anonfun$parmap$2(ThreadUtils.scala:373)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
	at scala.util.Success.$anonfun$map$1(Try.scala:255)
	at scala.util.Success.map(Try.scala:213)
```

### Does this PR introduce _any_ user-facing change?
Yes

### How was this patch tested?
Added UT to `LibSVMRelationSuite`.

### CODE REVIEW

**Comment by MaxGekk** [comment]
Thanks to @WeichenXu123 for the provided fix and the test. @HyukjinKwon @cloud-fan Please, review this PR. I am checking that the changes don't conflict with 3.0/2.4 branches and the bug exists there.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by MaxGekk** [comment]
Here are backports for:
- 3.0 https://github.com/apache/spark/pull/29675
- 2.4 https://github.com/apache/spark/pull/29678

FYI, the patch could be merged clearly but build will fail due to a missing function.

File: mllib/src/main/scala/org/apache/spark/ml/source/libsvm/LibSVMRelation.scala [hunk 1]
```diff
@@ -100,7 +100,7 @@ private[libsvm] class LibSVMFileFormat
         "though the input. If you know the number in advance, please specify it via " +
         "'numFeatures' option to avoid the extra scan.")
 
-      val paths = files.map(_.getPath.toUri.toString)
```
Comment (cloud-fan) [inline mllib/src/main/scala/org/apache/spark/ml/source/libsvm/LibSVMRelation.scala:103]: is this change related?

Comment (MaxGekk) [reply mllib/src/main/scala/org/apache/spark/ml/source/libsvm/LibSVMRelation.scala:103]: `getPath.toUri` do some escape on special characters, which break the case of path with glob metacharacters.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
